File: DataStructureJava/array/linkedList/LinkedList.java
public class LinkedList <E>{
    private class Node{
        public E e;
        public Node next;

        public Node(E e, Node next){
            this.e = e;
            this.next = next;
        }
        public Node(E e){
            this(e,null);
        }
        public Node(){
            this(null,null);
        }
        @Override
        public String toString(){
            return e.toString();
        }
    }
    private Node dummyhead;
    int size;
    public LinkedList(){
        dummyhead = new Node(null,null);
        size = 0;
    }

    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public void add(int index,E e){
        if (index < 0 || index > size)
            throw new IllegalArgumentException("add failed,Illegal index");
        Node prev = dummyhead;
        //从dummyhead开始遍历，找到index前面的哪一个节点
        for (int i = 0; i < index ; i++)
            prev = prev.next;
        prev.next = new Node(e,prev.next);
        size ++;
    }

    public void addFirst(E e){
        add(0,e);
    }

    public void addLast(E e){
        add(size,e);
    }

    public E get(int index){
        if (index <0 || index >= size)
            throw new IllegalArgumentException("get failed,Illegal index");
        Node cur = dummyhead.next;
        for (int i = 0; i < index;i ++)
            cur = cur.next;
        return cur.e;
    }

    public E getFirst() {
        return get(0);
    }

    public E getLast() {
        return get(size - 1);
    }

    public void set(int index, E e){
        if (index <0 || index >= size)
            throw new IllegalArgumentException("set failed,Illegal index");
        Node cur = dummyhead.next;
        for(int i = 0;i < index; i ++)
            cur = cur.next;
        cur.e = e;
    }

    public boolean contains(E e){
        Node cur = dummyhead.next;
        while(cur != null){
            if (cur.e.equals(e))
                return true;
            cur = cur.next;
        }
        return false;
    }

    public E remove(int index){
        if (index <0 || index >= size)
            throw new IllegalArgumentException("delete failed,Illegal index");
        Node prev = dummyhead;
        for(int i = 0;i < index; i ++)
            prev = prev.next;
        Node delNode = prev.next;
        prev.next = delNode.next;
        delNode.next = null;
        size --;
        return delNode.e;
    }

    public E removeFirst(){
        return remove(0);
    }

    public E removeLast(){
        return remove(size - 1);
    }

    public void removeElement(E e) {
        Node pre = dummyhead;
        while (pre.next != null) {
            if (pre.next.e.equals(e))
                break;
            pre = pre.next;
        }
        if (pre.next != null) {
            Node delNode = pre.next;
            pre.next = pre.next.next;
            delNode.next = null;
        }
    }


    @Override
    public String toString(){
        StringBuilder res = new StringBuilder();
        Node cur = dummyhead.next;
        while(cur != null){
            res.append(cur + "->");
            cur = cur.next;
        }
        res.append("null");
        return res.toString();
    }
}

File: data_structure_python/linked_list/linked_list.py

class ListNode(object):
    def __init__(self, val=None, next_=None):
        self.val = val
        self.next = next_

    def __str__(self):
        return str(self.val)

    def __repr__(self):
        return repr(self.val)


class LinkedList(object):
    def __init__(self):
        self.dummyHead = ListNode()
        self.size = 0

    def __str__(self):
        temp = self.dummyHead.next
        result = "["
        while temp is not None:
            result += str(temp.val)
            temp = temp.next
            if temp:
                result += ","
        result += "]"
        return result

    def __repr__(self):
        temp = self.dummyHead.next
        result = "["
        while temp is not None:
            result += repr(temp.val)
            temp = temp.next
        result += "]"
        return repr(result)

    def __len__(self):
        return self.size

    # if the list is empty
    def is_empty(self):
        return self.size == 0

    # add element
    def add(self, index, val):
        if index < 0 or index > self.size:
            raise IndexError("index is invalid")
        prev = self.dummyHead
        for i in range(index):
            prev = prev.next
        prev.next = ListNode(val, prev.next)
        self.size += 1

    # add element at front
    def add_first(self, val):
        self.add(0, val)

    # add element at the end
    def add_last(self, val):
        self.add(self.size, val)

    # get element value by index
    def get(self, index):
        if index < 0 or index >= self.size:
            raise IndexError("index is invalid")
        cur = self.dummyHead.next
        for i in range(index):
            cur = cur.next
        return cur.val

    # get first element
    def get_first(self):
        return self.get(0)

    # get last element
    def get_last(self):
        return self.get(self.size - 1)

    # return boolean if value is in the list
    def contains(self, val):
        cur = self.dummyHead.next
        while cur:
            if cur.val == val:
                return True
            cur = cur.next
        return False

    # update element value by index
    def set(self, index, val):
        if index < 0 or index >= self.size:
            raise IndexError("index is invalid")
        cur = self.dummyHead.next
        for i in range(index):
            cur = cur.next
        cur.val = val

    # remove element by index and return value
    def remove(self, index):
        if index < 0 or index >= self.size:
            raise IndexError("index is invalid")
        prev = self.dummyHead
        for i in range(index):
            prev = prev.next
        del_node = prev.next
        prev.next = del_node.next
        del_node.next = None
        self.size -= 1
        return del_node.val

    # remove first element
    def remove_first(self):
        return self.remove(0)

    # remove last element
    def remove_last(self):
        return self.remove(self.size - 1)

    # remove element by value
    def remove_element(self, val):
        pre = self.dummyHead
        cur = self.dummyHead.next
        while cur and cur.val != val:
            pre = cur
            cur = cur.next
        if cur.val == val:
            pre.next = cur.next
        self.size -= 1


if __name__ == '__main__':
    ll = LinkedList()
    ll.add_first(2)
    ll.add_last(3)
    print(ll)
    ll.remove(0)
    print(ll)
    # a = ll.contains(2)
    # print(a)